### DISCUSSION TITLE
Requirements for adding a Java Handler in Mkdocstrings?

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocstrings/mkdocstrings#703] [Q&A]
I recall having made a discussion or issue before on this, but I cannot find it...

I'm currently wondering, what requirements there would be in order to create a Handler for Mkdocstrings to process Java Documentation into Markdown files.

I preferably would want to use the source directly instead of first doing something like creating JSON (Which was the suggested aproach I remember) to the parse and handle.
I believe the main issue back then was the overall syntax?

Whatever the reason was exactly, I wanted to open this discussion, because I recently found out about Java 23 having Markdown support within their Javadocs system, meaning that it wouldn't use `{@link ...}` stuff anymore (Or not just that), but have a reference-link inspired syntax.

For an example, here is the common/old Javadoc syntax:
```java
/**
 * Lorem Ipsum dolor sit amet<br>
 * New line with a {@link com.example.Example Example project link}.
 *
 * <p>New Paragraph with an <a href="https://example.com">Example External Link</a>
 *
 * @param parameter1 The first parameter of this method
 * 
 * @return Something this would return...
 */
```
And here the new aproach using the Markdown Javadocs (Note the first line ends with two spaces):
```java
/// Lorem Ipsum dolor sit amet  
/// New line with a [Example project Link][com.example.Example].
/// 
/// New Paragraph with an [Example External Link](https://example.com)
///
/// @param parameter1 The first parameter of this method
/// 
/// @return Something this would return...
```

Not sure how much easier or difficult this would be to use now, but I think the overall markdown support seems significantly better than what was used before. It was noted in the feature explanation that CommonMark is being used in the background, so this should probably be kept in mind...

I'm still willing to try and work out a Handler, but the biggest issue has been so far that there wasn't any updated tool to create JSON from Javadoc comments, so maybe this would perhaps help a bit more? Idk...

**Comment by pawamoy** [comment]
The "official" issue is here: https://github.com/mkdocstrings/mkdocstrings/issues/99 (but no comments on it yet).

> I preferably would want to use the source directly instead of first doing something like creating JSON (Which was the suggested aproach I remember) to the parse and handle.

There are two approaches indeed:

- you collect data using a third-party or official tool that returns JSON (or any other format that can be loaded back into Python structures)
- you collect data directly in Python structures by writing or using a Java parser (for example Python bindings of treesitter) and walking the AST/CST of the sources

---

Good to know Markdown is better supported in Java now. In practice, for mkdocstrings, it means a Java handler won't have to bother converting docstrings since they're already written in Markdown :slightly_smiling_face: Well, looking at the second example, you'd probably still need to parse the syntax like `@param` to extract data in a structured way, unless you prefer delegating this to a Python-Markdown extension.

**Reply by Andre601** [reply]
Yeah, the JSON aproach is difficult. There is a tool (or as they are known for javadocs: Doclet) that can convert javadocs into JSON. But that one is outdated and uses API no longer used nor offered by Java. And there doesn't seem to be any other updated variants to use, whoch makes this a bit more difficult in the end.

----

The `@` tags are a little difficult to deal with, as they can be in different formats, and also handle vastly different depending on the types.

As an example, regarding formats, these two are both valid ways to format a tag:
```
@param parameter Description of parameter

@param parameter
       Description of parameter
```
Other cases can also be extra spaces between the tag and value.

Another thing is special behaviours for specific tags. For example:
```
@throws com.example.ExampleException When something bad happens

@see com.example.SeeExample
```
These would both treat the first value as a link with `@throw` also having a description. `@see` afaik can also be done this way:
```
@see [See Example][com.example.SeeExample]
```
This would override the displayed text itself.

So yeah, tags can be a bit difficult here to handle, so it would need some special handling there...

**Reply by pawamoy** [reply]
It's a bit like how we need to parse Google-style or Numpydoc-style docstrings for Python :) Not the most exciting thing to code, but definitely feasible ^^
